### ISSUE TITLE
Selection  will select as soon as you touch the window limits

### ISSUE THREAD

**Issue by Rmano** [xournalpp/xournalpp#875]
This is best explained by a gif: 

![recorded](https://user-images.githubusercontent.com/6414907/52732004-7170fd00-2fbf-11e9-9b5e-b1aaa819bb34.gif)

when you "touch" the window's border, it acts as if you ended the operation. The expected behavior, in my opinion, is that "pushing" will activate a scroll or that, at least, the selection is blocked --- the current behavior is quite non-standard.

**Comment by LittleHuba** [comment]
Yes, I think that is something that needs to be changed. Selection hitting window border should definitely scroll the view.

**Comment by LittleHuba** [comment]
Selection has to be rewritten anyway. We'll keep this in mind!
See #840

**Comment by Rmano** [comment]
Thanks!

**Comment by schaetzc** [comment]
Same goes for all drawing tools (for instance Arrow, Rectangle, ...). As soon as the mouse cursor leaves the canvas (the part of the window where the pages are displayed) the currently drawn object is written to the page and does not change anymore when entering the page.
When drawing you have to be very cautious not to touch the sidebar, toolbar, or leave the window.

Less severe but also annoying is that the object currently being drawn does not update when cursor has left the page but not the canvas.

![behavior of xournal++](https://user-images.githubusercontent.com/14296809/55220537-cc755100-5207-11e9-96a2-ac3763e5c6fb.gif)

It would be great if all drawing tools behaved the same way no matter the location of the pressed mouse cursor, just like in the old xournal:

![behavior of xournal](https://user-images.githubusercontent.com/14296809/55220553-d5662280-5207-11e9-99de-1baab2ac3494.gif)

Note that in xournal you could start drawing anywhere on the canvas, not necessarily on the page. which helped a lot drawing near the page borders.

![drawing outside the page in xournal](https://user-images.githubusercontent.com/14296809/55220908-a0a69b00-5208-11e9-9ff5-2d75943d86bd.gif)

**Comment by LittleHuba** [comment]
Will be fixed by #1070

**Comment by LittleHuba** [comment]
@schaetzc I'll have to check if your bug is already fixed by #1070. If it is not I'll handle it in a separate PR as #1070 is already big enough as it is.

**Comment by schaetzc** [comment]
@LittleHuba Thank you in advance!
I just tested [99a810e3](https://github.com/LittleHuba/xournalpp/commit/99a810e30f3328cd60e41b651c5f4537caafd0f8) from your fork's branch `fix_input`. Right now I didn't notice any changes for the problems mentioned in this issue. The selection tool still selects immediately when leaving the canvas. Lines currently being drawn only update while the cursor is on the page; leaving the canvas while drawing has the same effect as releasing the mouse button.

**Comment by LittleHuba** [comment]
There is currently a regression that prevents the new handling. I'll fix it asap.

**Comment by LittleHuba** [comment]
Lines won't be updated when drawing outside of the page in #1070 as this requires further work. But I have it still on my todos. It will be added once the new input system is merged in.

**Comment by LittleHuba** [comment]
@schaetzc Could you recheck if the behaviour is as expected concerning selections now?

**Comment by LittleHuba** [comment]
> Note that in xournal you could start drawing anywhere on the canvas, not necessarily on the page. which helped a lot drawing near the page borders.

This will not be possible to fix as the underlying logic of Xournal++ just not allows that.

**Comment by schaetzc** [comment]
I tried 11c32bbac37d8b9590bab282af289a6e628837c6. For me the selection still has the behavior Rmano reported in the first post of this issue.

I also noticed another problem in the old and new system: When drawing a freehand line, leaving the page and entering it again draws a straight line between the spots where you left and entered the page.
![leaving the page and entering it a again draws a straight line](https://user-images.githubusercontent.com/14296809/57358710-4a693800-7176-11e9-88c4-3684388d704e.gif)

**Comment by schaetzc** [comment]
Oops! Please excuse the confusion. I did not enable the new input system in the settings before – didn't know I had to.
I enabled the input system under `Edit`>`Preferences`>`Drawing Area`>`Input System`>`Enable new input system`. It's great! Thank you for your work.

Selection works as Rmano expected.
![select-arrows-outside](https://user-images.githubusercontent.com/14296809/57438778-fb3e0880-7244-11e9-9bfd-fde3542361fe.gif) &nbsp;&nbsp;&nbsp; ![select-arrows-outside2](https://user-images.githubusercontent.com/14296809/57438785-fda06280-7244-11e9-86a4-ef989f679d99.gif)

Drawn objects also update when the cursor leaves the page, canvas, or even window.
![draw-free-outside](https://user-images.githubusercontent.com/14296809/57439018-64be1700-7245-11e9-8f2d-27d79c0b052d.gif) &nbsp;&nbsp;&nbsp; ![draw-arrow-outside](https://user-images.githubusercontent.com/14296809/57439002-5c65dc00-7245-11e9-9ad0-cb93f2c88e69.gif)
![draw-free-outside2](https://user-images.githubusercontent.com/14296809/57439022-6687da80-7245-11e9-86e6-0b70da700c38.gif)

Pretty neat!
Just a bit of nitpicking: When the cursor (selection or drawing tool) leaves the canvas the page could scroll much faster. I think a very good automatic scrolling can be seen in most text editors where the scrolling speed increase when the cursor is further away from the window. The following animation shows this behavior for a [html text area](https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_textarea) in Firefox .
![auto-scroll-text](https://user-images.githubusercontent.com/14296809/57440127-b9fb2800-7247-11e9-99da-aa4e4c119a99.gif)

**Comment by LittleHuba** [comment]
> Oops! Please excuse the confusion. I did not enable the new input system in the settings before – didn't know I had to.
> I enabled the input system under `Edit`>`Preferences`>`Drawing Area`>`Input System`>`Enable new input system`. It's great! Thank you for your work.

Ahh I think I mentioned that somewhere but it may have been lost in all the messages.

> Selection works as Rmano expected.

Thats perfect!

> Drawn objects also update when the cursor leaves the page, canvas, or even window.

This should actually not happen for all tools, you tried, except the arrow... I'll have to revisit that.

> Just a bit of nitpicking: When the cursor (selection or drawing tool) leaves the canvas the page could scroll so much faster. I think a very good automatic scrolling can be seen in most text editors where the scrolling speed increases when the cursor is further away from the window. The following animation shows this behavior for a [html text area](https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_textarea) in Firefox .

This is already a TODO but I will add this in a follow-up PR. Currently it is more like a proof of concept...

**Comment by LittleHuba** [comment]
@schaetzc Could it be that you did not update to the newest version? There are things in your screenshot I have already fixed...

**Comment by schaetzc** [comment]
You are right, my last test was still on 11c32bbac37d8b9590bab282af289a6e628837c6.

I just checked out 928b7a618dc26d4375a9c740faaf0d304b6cf9f3 and re-compiled. Now the freehand drawing tools stop drawn when leaving the page. Bummer; I liked the scrolling behavior more.

![draw-outside](https://user-images.githubusercontent.com/14296809/57448169-04d16b80-7259-11e9-99d5-418bb830bceb.gif)

**Comment by LittleHuba** [comment]
@JJones780 @Technius @andreasb242 @morrolinux @peetCreative 
Some input on this please:
Should scrolling be enabled when pen reaches the border of the widget and should it be able to draw outside of the window?

**Comment by Rmano** [comment]
My opinion is that when drawing and touching the border of the widget, you should scroll if there is space on the page, and draw while scrolling _into_ the widget. I would not draw if the pen is outside the widget. 
If you reach the border of the page I am unsure about what to do... probably stop drawing and resume when re-entering the widget.

**Comment by LittleHuba** [comment]
I think this is problematic if you want to draw near the border. As moving the page is not really intuitive...

**Comment by schaetzc** [comment]
Yeah, I have to correct myself. Scrolling messes things up when drawing with a free hand pen tool.
I checked eight different programs.
Inkscape, gimp, libreoffice draw, pdf xchange, paint, paint 3d, and snip & sketch all allow you to draw outside of window but do not scroll the page. This feels pretty usable.
Only paint.net automatically scrolls (fast) when the cursor leaves the window. In my tests I wasn't able to draw something properly close to the window borders with this behavior.

Note that these observations only apply to the free hand drawing tools. Tools like the rectangle selection often scroll automatically, which seems fine.

**Comment by Rmano** [comment]
Yes, thinking a bit more I think that @schaetzc is right. I suppose that doing like gimp and inkscape is the best thing to do...

**Comment by Rmano** [comment]
@schaetzc sorry for the OT --- could you share which program are you using to capture your dynamic screenshots? Seems better than my `byzanz`...

**Comment by LittleHuba** [comment]
Okay so scrolling when reaching the border will be disallowed but you still want to be able to draw outside of the window.

@JJones780 @Technius @andreasb242 @morrolinux @peetCreative still interested in your input

**Comment by Technius** [comment]
I generally agree with @schaetzc, but I think the option to draw beyond the visible areas of the window should be enabled behind a checkbox somewhere. There are actually a few cases where the user may want to place a hard limit on the drawable area (e.g. to fit inside of a box).

**Comment by Technius** [comment]
@Rmano Check out [Peek](https://github.com/phw/peek). I use it all the time to make short gifs.

**Comment by schaetzc** [comment]
@Rmano I use the `Peek` (already linked by Technius). A really great tool.

**Comment by JJones780** [comment]
I agree with the result of Schaetzc's research of the other programs: automatic scrolling during drawing is bad and continued drawing outside the window is preferable.

 I would prefer that any drawing started on the page is linked to that page only ( but clipped to that page when that page isn't selected and shown even outside the page when it is selected ).  On the other hand... it's fun to scribble over multiple pages!  Either way is acceptable.

I also found peek to be handy for creating animated gifs.

**Comment by LittleHuba** [comment]
> ( but clipped to that page when that page isn't selected and shown even outside the page when it is selected )

This is not possible with how Xournal++ renders the page content. Content on a page needs to be within the bounds of the page or you get graphical glitches.
